### ISSUE TITLE
sys.path clash for pyenv-managed python and system python

### ISSUE THREAD

**Issue by krautcat** [pyenv/pyenv#1606]
### Prerequisite
<details>
  <summary>Here</summary>

> * [x] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).

Not the build problem.

> * [x] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.

Didn't find, but think it could be connected with issues I'll mention later.

> * [x] Make sure you are not asking us to help solving your specific issue.

>  * GitHub issues is opened mainly for development purposes. If you want to ask someone to help solving your problem, go to some community site like [Gitter](https://gitter.im/yyuu/pyenv), [StackOverflow](https://stackoverflow.com/questions/tagged/pyenv), etc.

It's not this case.

> * [x] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
>  * Please refer to the package documentation for the installation issues, etc.

Not the package manager issues.

> * [x] Make sure your problem is not derived from plugins.
>   * This repository is maintaining `pyenv` and the default `python-build` plugin only. Please refrain from reporting issues of other plugins here.

I'll mention one plugin because it get its job right.
</details>

### Description

<details>
  <summary>Here</summary>

- [x] Platform information: Ubuntu 18.04.3
- [x] OS architecture: amd64
- [x] pyenv version: pyenv 1.2.9-152-g5b009e87
- [x] Python version: compatible (major, minor) with system 
- [x] C Compiler information: not relevant 
- [x] Please attach verbose build log as gist: not relevant

</details>

### TL;DR

It seems pyenv tends to include system python's user site-packages path to `sys.path` of pyenv-managed python with the same major and minor version as system one. 

I don't know whether it's expected behaviour or not but it's confusing. I expect I can use ipython from user's site-packages or, what's better, install ipython for pyenv-managed python version.

### Full description

Well, there's strange behaviour maybe connected with this issues #1475, #1204. Maybe it's of clash my system versin python and pythons managed by pyenv.

I installed ipython for my virtualenvs.
<details>
  <summary>It seems path where bin located is good.</summary>

```
gerrit-api.py.3.6.0 ❯ pyenv activate 3.6.0/envs/gerrit-api.py.3.6.0
pyenv-virtualenv: version `3.6.0/envs/gerrit-api.py.3.6.0' is already activated

gerrit-api.py.3.6.0 ❯ pyenv which ipython
/home/krautcat/.local/env/pyenv/versions/3.6.0/envs/gerrit-api.py.3.6.0/bin/ipython

gerrit-api.py.3.6.0 ❯ pyenv activate 3.6.1/envs/bsa-server-updater.py.3.6.1

bsa-server-updater.py.3.6.1 ❯ pyenv which ipython
/home/krautcat/.local/env/pyenv/versions/3.6.1/envs/bsa-server-updater.py.3.6.1/bin/ipython
```
</details>

<details>
  <summary>Path to bins from info via pip seems to be good.</summary>

```
bsa-server-updater.py.3.6.1 ❯ unbuffer pip show -f ipython | head -n 15
Name: ipython
Version: 7.14.0
Summary: IPython: Productive Interactive Computing
Home-page: https://ipython.org
Author: The IPython Development Team
Author-email: ipython-dev@python.org
License: BSD
Location: /home/krautcat/.local/env/pyenv/versions/3.6.1/envs/bsa-server-updater.py.3.6.1/lib/python3.6/site-packages
Requires: backcall, jedi, setuptools, pygments, pexpect, traitlets, decorator, pickleshare, prompt-toolkit
Files:
  ../../../bin/iptest
  ../../../bin/iptest3
  ../../../bin/ipython
  ../../../bin/ipython3
  ../../../share/man/man1/ipython.1.gz
```
</details>

<details>
  <summary>I also installed ipython for system python. It's important in my issue I think.</summary>

```
❯ pyenv shell system

❯ pyenv which python3
/usr/bin/python3

❯ python3 -c "import sys; print ('\n'.join(sys.path))"

/usr/lib/python36.zip
/usr/lib/python3.6
/usr/lib/python3.6/lib-dynload
/home/krautcat/.local/lib/python3.6/site-packages
/usr/local/lib/python3.6/dist-packages
/usr/lib/python3/dist-packages
```
</details>

<details>
  <summary>Paths listed by pip seems to be good too.</summary>

```
❯ unbuffer pip3 show -f ipython | head -n 15
Name: ipython
Version: 7.9.0
Summary: IPython: Productive Interactive Computing
Home-page: https://ipython.org
Author: The IPython Development Team
Author-email: ipython-dev@python.org
License: BSD
Location: /home/krautcat/.local/lib/python3.6/site-packages
Requires: decorator, backcall, pickleshare, prompt-toolkit, pexpect, pygments, setuptools, jedi, traitlets
Required-by: ipdb
Files:
  ../../../bin/iptest
  ../../../bin/iptest3
  ../../../bin/ipython
  ../../../bin/ipython3
```
</details>

Ok, let's try to install ipython for pyenv managed versions of python whose major and minor versions are the same with system python.

<details>
  <summary>I have even three such pythons.</summary>

```
❯ pyenv versions | grep -iP "^\s*3\.6\.\d{1,2}\s*$"
  3.6.0
  3.6.1
  3.6.9
```
</details>

<details>
   <summary>Go and install ipython for my pyenv-managed 3.6.1.</summary>

```
❯ pyenv shell 3.6.1

❯ pip install ipython
Requirement already satisfied: ipython in /home/krautcat/.local/lib/python3.6/site-packages (7.9.0)
... some lines more ...
```
</details>

Wait... Why doest it claim requirement is already satisfied? Well, lets try to point the `--target` path for ipython.

<details>
  <summary>It seems pip installs ipython... Let's pretend warnings are not essential here.</summary>

 ```
❯ pip install --target /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages ipython
Collecting ipython
  Using cached ipython-7.14.0-py3-none-any.whl (782 kB)
Collecting setuptools>=18.5
  Using cached setuptools-46.1.3-py3-none-any.whl (582 kB)
Collecting decorator
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Processing /home/krautcat/.cache/pip/wheels/98/b0/dd/29e28ff615af3dda4c67cab719dd51357597eabff926976b45/backcall-0.1.0-cp36-none-any.whl
Collecting traitlets>=4.2
  Using cached traitlets-4.3.3-py2.py3-none-any.whl (75 kB)
Collecting jedi>=0.10
  Using cached jedi-0.17.0-py2.py3-none-any.whl (1.1 MB)
Collecting pickleshare
  Using cached pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)
Collecting pexpect; sys_platform != "win32"
  Using cached pexpect-4.8.0-py2.py3-none-any.whl (59 kB)
Collecting prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0
  Using cached prompt_toolkit-3.0.5-py3-none-any.whl (351 kB)
Collecting pygments
  Using cached Pygments-2.6.1-py3-none-any.whl (914 kB)
Collecting six
  Using cached six-1.14.0-py2.py3-none-any.whl (10 kB)
Collecting ipython-genutils
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
Collecting parso>=0.7.0
  Using cached parso-0.7.0-py2.py3-none-any.whl (100 kB)
Collecting ptyprocess>=0.5
  Using cached ptyprocess-0.6.0-py2.py3-none-any.whl (39 kB)
Collecting wcwidth
  Using cached wcwidth-0.1.9-py2.py3-none-any.whl (19 kB)
Installing collected packages: setuptools, decorator, backcall, six, ipython-genutils, traitlets, parso, jedi, pickleshare, ptyprocess, pexpect, wcwidth, prompt-toolkit, pygments, ipython
Successfully installed backcall-0.1.0 decorator-4.4.2 ipython-7.14.0 ipython-genutils-0.2.0 jedi-0.17.0 parso-0.7.0 pexpect-4.8.0 pickleshare-0.7.5 prompt-toolkit-3.0.5 ptyprocess-0.6.0 pygments-2.6.1 setuptools-46.1.3 six-1.14.0 traitlets-4.3.3 wcwidth-0.1.9
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/pexpect already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/Pygments-2.6.1.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/six.py already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/ptyprocess already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/pygments already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/ptyprocess-0.6.0.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/__pycache__ already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/pkg_resources already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/easy_install.py already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/six-1.14.0.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/pexpect-4.8.0.dist-info already exists. Specify --upgrade to force replacement.
WARNING: Target directory /home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages/setuptools already exists. Specify --upgrade to force replacement.
```
</details>

Then try to run ipython. 

<details>
  <summary>It fails.</summary>

```
❯ ipython
pyenv: ipython: command not found

The `ipython' command exists in these Python versions:
... some versions and virtualenvs ...
```
</details>

Let's examine bin path of pyenv managed python version.
<details>
  <summary>No <code>ipython</code> here.</summary>

```
❯ tree -L 1 /home/krautcat/.local/env/pyenv/versions/3.6.1/bin
/home/krautcat/.local/env/pyenv/versions/3.6.1/bin
├── 2to3 -> 2to3-3.6
├── 2to3-3.6
├── chardetect
├── easy_install -> easy_install-3.6
├── easy_install-3.6
├── idle -> idle3.6
├── idle3 -> idle3.6
├── idle3.6
├── pip
...
```
</details>

So why does it fail? Why doesn't pip install ipython for my pyenv-managed pythons unfortunately compatible with system? I printed sys.path for pyenv-managed python version and maybe it can be the origin of trouble.

<details>
  <summary>And what tell us sys.path...</summary>

```
❯ pyenv version
3.6.1 (set by PYENV_VERSION environment variable)

❯ python -c "import sys; print ('\n'.join(sys.path))"

/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python36.zip
/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6
/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/lib-dynload
/home/krautcat/.local/lib/python3.6/site-packages
/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/site-packages
```
</details>

And there's some more `sys.path`s for different versions and virtualenvs. It's obvious that it shouldn't include user's site-package directory for python with non-compatible minor version. But it's interesting for me that virtualenv handles `sys.path` correctly.

<details>
  <summary>Different <code>sys.path</code>s. They don't include user's site-package directory even for virtualenv for python compatible with system one.</summary>

```
❯ pyenv shell 3.7.2

❯ python -c "import sys; print ('\n'.join(sys.path))"

/home/krautcat/.local/env/pyenv/versions/3.7.2/lib/python37.zip
/home/krautcat/.local/env/pyenv/versions/3.7.2/lib/python3.7
/home/krautcat/.local/env/pyenv/versions/3.7.2/lib/python3.7/lib-dynload
/home/krautcat/.local/env/pyenv/versions/3.7.2/lib/python3.7/site-packages

❯ pyenv shell 3.5.2

❯ python -c "import sys; print ('\n'.join(sys.path))"

/home/krautcat/.local/env/pyenv/versions/3.5.2/lib/python35.zip
/home/krautcat/.local/env/pyenv/versions/3.5.2/lib/python3.5
/home/krautcat/.local/env/pyenv/versions/3.5.2/lib/python3.5/plat-linux
/home/krautcat/.local/env/pyenv/versions/3.5.2/lib/python3.5/lib-dynload
/home/krautcat/.local/env/pyenv/versions/3.5.2/lib/python3.5/site-packages

❯ pyenv activate 3.6.1/envs/bsa-server-updater.py.3.6.1

bsa-server-updater.py.3.6.1 ❯ python -c "import sys; print ('\n'.join(sys.path))"

/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python36.zip
/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6
/home/krautcat/.local/env/pyenv/versions/3.6.1/lib/python3.6/lib-dynload
/home/krautcat/.local/env/pyenv/versions/3.6.1/envs/bsa-server-updater.py.3.6.1/lib/python3.6/site-packages
```
</details>

**Comment by wadkar** [comment]
Long time ago, I got bitten by something like this where I suspected a system pip install was messing up with my project install. I make it a point to not touch system python at all.

This is a very good bug report. And as of now I have no idea what could be wrong here except what you've deduced so far (i.e. something's up with how `sys.path` is handled)

/me 👀

**Comment by UnoDosThree** [comment]
The issue is that pyenv leaves the user site enabled for its managed pythons, which adds the user site path to sys.path.  Also the user site path is added before managed python's site path.

We would see the behaviour that @krautcat is running into, if we try to install a package into any NON-managed python (system or otherwise), and then try to install that same package into a pyenv managed python.

The reason why @krautcat's virtualenvs are handling the sys.path correctly, is that the virtualenvs have a pyvenv.cfg with the line:

include-system-site-packages = false

This disables the user site thereby omitting the user site path from sys.path.  You can check this by activating a venv and then running the command

python -m site

You should see a line "ENABLE_USER_SITE: False" and you won't see the user site path in the sys.path.

**Comment by native-api** [comment]
WFM now so must've been fixed since.
